src/EditorFeatures/Core/ExternalAccess/IntelliCode/Api/IIntentProcessor.cs

```diff
@@ -0,0 +1,96 @@
+        /// For an input intent, computes the edits required to apply that intent and returns them.
+        /// </summary>
+        /// <param name="intentRequestContext">the intents with the context in which the intent was found.</param>
+        /// <returns>the edits that should be applied to the current snapshot.</returns>
+        Task<ImmutableArray<IntentResult>> ComputeEditsAsync(IntentRequestContext intentRequestContext, CancellationToken cancellationToken);
+    }
+
+    /// <summary>
+    /// Defines the data needed to compute the code action edits from an intent.
+    /// </summary>
+    internal struct IntentRequestContext
+    {
+        /// <summary>
+        /// The intent name.  <see cref="WellKnownIntents"/> contains all intents roslyn knows how to handle.
+        /// </summary>
+        public string IntentName { get; }
+
+        /// <summary>
+        /// JSON formatted data specific to the intent that must be deserialized into the appropriate object.
+        /// </summary>
+        public string? IntentData { get; }
+
+        /// <summary>
+        /// The text snapshot and selection when <see cref="IIntentProcessor.ComputeEditsAsync"/>
+        /// was called to compute the text edits and against which the resulting text edits will be calculated.
+        /// </summary>
+        public SnapshotSpan CurrentSnapshotSpan { get; }
+
+        /// <summary>
+        /// The text edits that should be applied to the <see cref="CurrentSnapshotSpan"/> to calculate
+        /// a prior text snapshot before the intent happened.  The snapshot is used to calculate the actions.
+        /// </summary>
+        public ImmutableArray<TextChange> PriorTextEdits { get; }
+
+        /// <summary>
+        /// The caret position / selection in the snapshot calculated by applying
+        /// <see cref="PriorTextEdits"/> to the <see cref="CurrentSnapshotSpan"/>
+        /// </summary>
+        public TextSpan PriorSelection { get; }
+
+        public IntentRequestContext(string intentName, SnapshotSpan currentSnapshotSpan, ImmutableArray<TextChange> textEditsToPrior, TextSpan priorSelection, string? intentData)
+        {
+            IntentName = intentName ?? throw new ArgumentNullException(nameof(intentName));
+            IntentData = intentData;
+            CurrentSnapshotSpan = currentSnapshotSpan;
+            PriorTextEdits = textEditsToPrior;
+            PriorSelection = priorSelection;
+        }
+    }
+
+    /// <summary>
+    /// Defines the text changes needed to apply an intent.
+    /// </summary>
+    internal struct IntentResult
+    {
+        /// <summary>
+        /// The text changes that should be applied to the <see cref="IntentRequestContext.CurrentSnapshotSpan"/>
+        /// </summary>
+        public readonly ImmutableArray<TextChange> TextChanges;
+
+        /// <summary>
+        /// The title associated with this intent result.
+        /// </summary>
+        public readonly string Title;
```

Comment:
nit: sort this above TextChanges.
nit: you use fields here, but properties in teh above object.  can you be consistent for both?

since this will be a public API, properties are almost certainly more appropraite as they give more flexibility in teh future to change things. (we'd also likely use itnerfaces here in the future)

Reply:
will move, but what are the sorting rules here?  alphabetical, based on type?